src/codecs/dds/decoder.rs

```diff
@@ -0,0 +1,1112 @@
+            Self::R16_FLOAT => decode_R16_FLOAT(r, size, buf),
+            Self::R16G16_FLOAT => decode_R16G16_FLOAT(r, size, buf),
+            Self::R16G16B16A16_FLOAT => decode_R16G16B16A16_FLOAT(r, size, buf),
+            Self::R32_FLOAT => decode_R32_FLOAT(r, size, buf),
+            Self::R32G32_FLOAT => decode_R32G32_FLOAT(r, size, buf),
+            Self::R32G32B32_FLOAT => decode_R32G32B32_FLOAT(r, size, buf),
+            Self::R32G32B32A32_FLOAT => decode_R32G32B32A32_FLOAT(r, size, buf),
+            Self::R10G10B10_XR_BIAS_A2_UNORM => decode_R10G10B10_XR_BIAS_A2_UNORM(r, size, buf),
+
+            // sub-sampled formats
+            Self::R8G8_B8G8_UNORM => decode_R8G8_B8G8_UNORM(r, size, buf),
+            Self::G8R8_G8B8_UNORM => decode_G8R8_G8B8_UNORM(r, size, buf),
+
+            // block compression formats
+            Self::BC1_ALPHA_UNORM => decode_BC1_ALPHA_UNORM(r, size, buf),
+            Self::BC2_UNORM => decode_BC2_UNORM(r, size, buf),
+            Self::BC3_UNORM => decode_BC3_UNORM(r, size, buf),
+            Self::BC4_UNORM => decode_BC4_UNORM(r, size, buf),
+            Self::BC4_SNORM => decode_BC4_SNORM(r, size, buf),
+            Self::BC5_UNORM => decode_BC5_UNORM(r, size, buf),
+            Self::BC5_SNORM => decode_BC5_SNORM(r, size, buf),
+            Self::BC7_UNORM => decode_BC7_UNORM(r, size, buf),
+        }
+    }
+}
+
+pub(crate) struct DX10Decoder<R: Read> {
+    pub(crate) width: u32,
+    pub(crate) height: u32,
+    pub(crate) format: SupportedFormat,
+    pub(crate) cube: bool,
+    /// The number of mipmap levels excluding the top level.
+    pub(crate) mip_count: u8,
+    pub(crate) inner: R,
+}
+
+impl<R: Read> DX10Decoder<R> {
+    pub(crate) fn get_cube_size(width: u32, height: u32) -> Option<(u32, u32)> {
+        Some((width.checked_mul(4)?, height.checked_mul(3)?))
+    }
+
+    fn read_surface(&mut self, buf: &mut [u8]) -> ImageResult<()> {
+        let r: &mut dyn Read = &mut self.inner;
+        let size = Size {
+            width: self.width as usize,
+            height: self.height as usize,
+        };
+
+        self.format.read_surface(r, size, buf)
+    }
+
+    fn read_cube(&mut self, cube_buf: &mut [u8]) -> ImageResult<()> {
+        // zero-out the buffer to make it transparent
+        cube_buf.fill(0);
+
+        let w = self.width as usize;
+        let h = self.height as usize;
+
+        let surface_color = self.format.color_type();
+
+        let surface_channels = surface_color.channel_count() as usize;
+
+        let face_offsets =
+            [(2, 1), (0, 1), (1, 0), (1, 2), (1, 1), (3, 1)].map(|(x, y)| (x * w, y * h));
+
+        match surface_color {
+            ColorType::L8 | ColorType::La8 | ColorType::Rgb8 | ColorType::Rgba8 => {
+                todo!()
+            }
+            ColorType::L16 | ColorType::La16 | ColorType::Rgb16 | ColorType::Rgba16 => {
+                let buf: &mut [u16] = bytemuck::cast_slice_mut(cube_buf);
```

Comment:
I was under the assumption that the buffer passed to `ImageDecoder::read_image` was required to be aligned to the element type of the decoder's color type. Good to know that this isn't the case. I'll review all my uses casting.

Reply:
Can you describe how you arrived at your assumption? Honestly asking since we should make sure this information ends up there and it's hard to look at documentation as if from a fresh pair of eyes.

For reference, it's currently here: https://docs.rs/image/latest/image/trait.ImageDecoder.html#tymethod.read_image

> Although not required, for certain color types callers may want to pass buffers which are aligned to 2 or 4 byte boundaries […]

Maybe the wording here is also confusing. Hm.

Reply:
I think, I just skimmed through the documentation and the "Although not required" was on a different line in my IDE. Because this sentence makes it pretty clear that the buffer isn't guaranteed to be aligned.

Reply:
I think I fixed it everywhere. I made it so that the only cast that can fail uses `try_cast_slice_mut`. All other slice casts cast between `&[T]` <-> `&[[T; N]]`. Since arrays have the same alignment as their item type, these casts cannot fail (assuming that the source slice has the right number of elements, but the implementation makes sure of that).